cmds/field/field.go

```diff
@@ -0,0 +1,366 @@
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"unicode"
+	"unicode/utf8"
+)
+
+type frange struct {
+	begin int
+	end   int
+}
+
+const nf = 0x7FFFFFFF
+
+var (
+	zflag      = flag.Bool("0", false, "use the NUL character ('\\0') as output separator")
+	eflag      = flag.Bool("e", false, "")
+	Eflag      = flag.Bool("E", false, "")
+	insepflag  = flag.String("F", "[ \t\v\r]+", "Input separator characters (regular expression)")
+	outsepflag = flag.String("O", " ", "Output separater (string)")
+)
+
+func main() {
+	flag.Parse()
+
+	fstate := make(map[string]bool)
+	flag.Visit(func(f *flag.Flag) { fstate[f.Name] = true })
+	if fstate["e"] && fstate["E"] {
+		fatal("flag conflict: -e and -E are mutually exclusive")
+	}
+	if fstate["0"] && fstate["O"] {
+		fatal("flag conflict: -O and -0 are mutually exclusive")
+	}
+
+	outsep := *outsepflag
+	insep := *insepflag
+	collapse := shouldcollapse(insep)
+	delim, re_err := regexp.Compile(insep)
+	if re_err != nil {
+		fatal("Delimiter regexp failed to parse: %v", re_err)
+	}
+
+	if flag.NArg() == 0 {
+		fatal("Range specifier missing")
+	}
+	rv := parseranges(flag.Arg(0))
+
+	if flag.NArg() == 1 {
+		process(os.Stdin, rv, delim, outsep, collapse)
+		return
+	}
+	for i := 1; i < flag.NArg(); i++ {
+		filename := flag.Arg(i)
+		if filename == "-" {
+			process(os.Stdin, rv, delim, outsep, collapse)
+			continue
+		}
+		file, err := os.Open(filename)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Cannot open:", err)
+			continue
+		}
+		process(file, rv, delim, outsep, collapse)
+		file.Close()
+	}
+}
+
+func shouldcollapse(s string) bool {
+	if *eflag {
+		return false
+	}
+	if *Eflag {
+		return true
+	}
+	l := utf8.RuneCountInString(s)
+	r, _ := utf8.DecodeRuneInString(s)
+	return l > 1 && (l != 2 || r != '\\')
+}
+
+// The field selection syntax is:
+//
+// ranges := range [[delim] range]
+// range := field | NUM '-' [field]
+// field := NUM | NF
+// delim := ws+ | '|' | ','
+// ws := c such that `isspace(c)` is true.
+// NF := 'NF' | 'N'
+// (Numbers can be negative)
+
+func parseranges(input string) []frange {
+	rs := []frange{}
+	lex := &lexer{input: input}
+	if input == "" {
+		fatal("Empty field range")
+	}
+	lex.next()
+	for {
+		for typ := lex.peektype(); typ == TokSpace; typ = lex.peektype() {
```

Comment:
Though it occurs to me that whitespace is accumulated into a single token, so instead of a loop, one should be able to get away with a single peektype and next.